esp-hal/src/lib.rs

```diff
@@ -644,3 +644,31 @@ macro_rules! before_snippet {
         ))
     };
 }
+
+use crate::{
+    clock::{ClockControl, Clocks, CpuClock},
+    peripherals::Peripherals,
+};
+
+/// System components.
+pub struct System {
+    /// The available peripheral instances.
+    pub peripherals: Peripherals,
+
+    /// The configured clocks.
+    pub clocks: Clocks<'static>,
+
+    /// The available software interrupts.
+    pub software_interrupt_control: crate::system::SoftwareInterruptControl,
+}
+
+/// Initialize the system.
+pub fn init(clock_config: CpuClock) -> System {
```

Comment:
I suppose this means no more changing clock frequency after initialization.

I haven't tried this myself but is it a valid use case for someone to temporarily drop the clock frequency to save power?

Reply:
I don't think this has ever been possible (in an intended way) previousy, at least how the documentation is written seems to imply this.

I'm aware that this may be more limiting than what we want, but resolving it properly is not trivial. For one `esp-wifi` can't be disabled, but I'm fairly sure it would react to changing clocks quite poorly.

Reply:
Ahh fair point, that would need https://github.com/esp-rs/esp-hal/issues/1604 . Nvm then, I guess we can leave this problem for later.

Reply:
I think we deliberately opted for fixed clocks - IIRC the very first ESP32-HAL tried to deal with dynamic clocks (e.g. triggering drivers to re-calculate the clock dividers) but that was a huge PITA - also not sure that ever reliably worked

Probably for power save putting the chip to light or deep sleep is a better option anyways

Reply:
Ah I didn't know about that. I think supporting dynamic clocks means users have to be (somewhat) on their own when it comes to frequency calculation for each driver. Trying to make it automatically update peripheral clock settings would definitely suck and probably wouldn't be the way to go.

Light sleep is a viable option when there's nothing else to do but what if there's still work to do but you don't need much power to do it?

The current code in repo forces the invariant that, as long as a clock sensitive driver exists then the global clocks can't be changed. (The doc doesn't quite match that but I think that's what the code is doing) 
At least then users could just recreate the clock sensitive drivers with the new clock settings when it's updated. 

For the case of esp-wifi, the immediate fix there is to accept a 'static clock in the constructor, then that enforces that when WiFi driver is initialized, the clocks can no longer be changed, ever. At least until the WiFi driver is allowed to be deinitialised.

I currently don't see nor understand the motive for throwing away the current clock setting code when it's doing an alright job. Throwing it away now would just make it really difficult (or close to impossible tbh) to bring back and improve. Unless there are actually problems with it, can we just keep it around?